okhttp/src/main/java/io/grpc/okhttp/OkHttpChannelBuilder.java

```diff
@@ -418,6 +466,99 @@ SSLSocketFactory createSslSocketFactory() {
     }
   }
 
+  private static final EnumSet<TlsChannelCredentials.Feature> understoodTlsFeatures =
+      EnumSet.noneOf(TlsChannelCredentials.Feature.class);
+
+  static SslSocketFactoryResult sslSocketFactoryFrom(ChannelCredentials creds) {
+    if (creds instanceof TlsChannelCredentials) {
+      TlsChannelCredentials tlsCreds = (TlsChannelCredentials) creds;
+      Set<TlsChannelCredentials.Feature> incomprehensible =
+          tlsCreds.incomprehensible(understoodTlsFeatures);
+      if (!incomprehensible.isEmpty()) {
+        return SslSocketFactoryResult.error(
+            "TLS features not understood: " + incomprehensible);
+      }
+      SSLSocketFactory sslSocketFactory;
+      try {
+        SSLContext sslContext = SSLContext.getInstance("Default", Platform.get().getProvider());
+        sslSocketFactory = sslContext.getSocketFactory();
+      } catch (GeneralSecurityException gse) {
+        throw new RuntimeException("TLS Provider failure", gse);
+      }
+      return SslSocketFactoryResult.factory(sslSocketFactory);
+
+    } else if (creds instanceof InsecureChannelCredentials) {
+      return SslSocketFactoryResult.plaintext();
+
+    } else if (creds instanceof CompositeChannelCredentials) {
+      CompositeChannelCredentials compCreds = (CompositeChannelCredentials) creds;
+      return sslSocketFactoryFrom(compCreds.getChannelCredentials())
+          .withCallCredentials(compCreds.getCallCredentials());
+
+    } else if (creds instanceof SslSocketFactoryChannelCredentials.ChannelCredentials) {
+      SslSocketFactoryChannelCredentials.ChannelCredentials factoryCreds =
+          (SslSocketFactoryChannelCredentials.ChannelCredentials) creds;
+      return SslSocketFactoryResult.factory(factoryCreds.getFactory());
+
+    } else if (creds instanceof ChoiceChannelCredentials) {
+      ChoiceChannelCredentials choiceCreds = (ChoiceChannelCredentials) creds;
+      StringBuilder error = new StringBuilder();
+      for (ChannelCredentials innerCreds : choiceCreds.getCredentialsList()) {
+        SslSocketFactoryResult result = sslSocketFactoryFrom(innerCreds);
+        if (result.error == null) {
+          return result;
+        }
+        error.append(", ");
+        error.append(result.error);
+      }
+      return SslSocketFactoryResult.error(error.substring(2));
+
+    } else {
+      return SslSocketFactoryResult.error(
+          "Unsupported credential type: " + creds.getClass().getName());
+    }
+  }
+
+  static final class SslSocketFactoryResult {
+    /** {@code null} implies plaintext if {@code error == null}. */
+    public final SSLSocketFactory factory;
+    public final CallCredentials callCredentials;
+    public final String error;
```

Comment:
These 3 don't need to be public and can be marked `@Nullable`

Reply:
I use public when it is public from an API perspective (the consumer is expected to use these fields). So public is appropriate here. I only use Nullable for consistency and sometimes for public APIs. This is a new internal class.

Reply:
okay